### PR TITLE
Fix negative axes reductions.

### LINKED ISSUE #117: Reduce operations raise error for negative axis values
Reduce operations like `.sum()`, `.prod()` etc. raise exceptions when you pass a negative `axis` value:

```python
import sparse
sparse.random((40, 50)).sum(0)  # works
sparse.random((40, 50)).sum(1)  # works
sparse.random((40, 50)).sum(-1)  # fails, should be equivalent to .sum(1)
```

The traceback I get:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "proj_dir/sparse/sparse/coo.py", line 1038, in prod
    return self.reduce(np.multiply, axis=axis, keepdims=keepdims, dtype=dtype)
  File "proj_dir/sparse/sparse/coo.py", line 771, in reduce
    a = self.transpose(neg_axis + axis)
  File "proj_dir/sparse/sparse/coo.py", line 1099, in transpose
    raise ValueError("repeated axis in transpose")
ValueError: repeated axis in transpose
```

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #117

### CODE REVIEW

File: sparse/coo.py [hunk 1]
```diff
@@ -712,13 +712,15 @@ def reduce(self, method, axis=(0,), keepdims=False, **kwargs):
...
+        axis = tuple(a if a >= 0 else a + self.ndim for a in axis)
+
         if set(axis) == set(range(self.ndim)):
             result = method.reduce(self.data, **kwargs)
             if self.nnz != self.size:
                 result = method(result, _zero_of_dtype(self.dtype)[()], **kwargs)
         else:
             axis = tuple(axis)
-            neg_axis = tuple(ax for ax in range(self.ndim) if ax not in axis)
+            neg_axis = tuple(ax for ax in range(self.ndim) if ax not in set(axis))
```
Comment (mrocklin) [inline sparse/coo.py]: Why this change?  I'm not entirely sure of Python semantics here but I suspect that you're creating a new set every time.

File: sparse/coo.py [hunk 2]
```diff
@@ -712,13 +712,15 @@ def reduce(self, method, axis=(0,), keepdims=False, **kwargs):
         if not isinstance(axis, tuple):
             axis = (axis,)
 
+        axis = tuple(a if a >= 0 else a + self.ndim for a in axis)
```
Comment (mrocklin) [inline sparse/coo.py]: Are there other uses of `axis=` where this should be done but is not yet currently done?

File: sparse/coo.py (hunk 1 above)
Comment (hameerabbasi) [reply sparse/coo.py]: Makes it `O(ndim)` rather than `O(ndim^2)`. I ran a quick test to see if this was indeed the case:

```python
>>> class A:
...     def get_set(self):
...         print("Called")
...         return set(range(5))
...     
>>> tuple(i for i in A().get_set())
Called
(0, 1, 2, 3, 4)
```

File: sparse/coo.py (hunk 2 above)
Comment (hameerabbasi) [reply sparse/coo.py]: With a short "find all" of `axis`, I'm inclined to say no. It does appear in `tensordot`, `concatenate` and `stack` but those already have this.

File: sparse/coo.py (hunk 1 above)
Comment (mrocklin) [reply sparse/coo.py]: This doesn't really matter either way.   Asymptotic arguments don't really carry significance when n is likely to be very small.


```python
In [1]: axis = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

In [2]: ndim = 20

In [3]: %timeit tuple(ax for ax in range(ndim) if ax not in axis)
The slowest run took 6.44 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 4.03 µs per loop
```

Happy either way though.  I'm glad to learn that Python handles things in that order.
